src/controllers/settings_controller.js

```diff
@@ -228,24 +229,108 @@ SettingsController.post('/skiptunnel', async (request, response) => {
+      tunnelEndpoint = `https://${result.name}.
+      ${config.get('ssltunnel.domain')}`;
     } else {
-      response.status(404).end();
+      tunnelEndpoint = 'Not set.';
+    }
+
+    // Find out our default local DNS name Check for a previous name in the
+    // DB, if that does not exist use the default.
+    try {
+      mDNSstate = await Settings.get('multicastDNSstate');
+      localDomain = await Settings.get('localDNSname');
+    } catch (err) {
+      // Catch this DB error. Since we don't know what state the mDNS process
+      // should be in make sure it's off
+      console.error(`Error getting DB entry for multicast from the DB: ${err}`);
+      localDomain = config.get(
+        'settings.defaults.domain.localControl.mdnsServiceDomain');
     }
+
+    console.log(`Tunnel name is set to: ${tunnelEndpoint}`);
+    console.log(`Local mDNS Service Domain Name is: ${localDomain}`);
+    const localDomainSettings = {localDomain: localDomain,
+                                 mDNSstate: mDNSstate,
+                                 tunnelDomain: tunnelEndpoint};
+    response.send(localDomainSettings);
+    response.status(200).end();
   } catch (e) {
-    console.error('Failed to retrieve tunneltoken setting');
+    console.error('Failed to retrieve default settings for ' +
+      'tunneltoken or local service discovery setting');
     console.error(e);
     response.status(400).send(e);
   }
 });
 
+/* This is responsible for controlling dynamically the local domain name
+    settings (via mDNS) and changing or updating mozilla tunnel endpoints.
+    The /domain endpoint is invoked from:
+    MainMenu -> Settings -> Doamin
+
+    JSON data: { local: {
+                    multicastDNSstate: boolean,
+                    localDNSname: string, - e.g. MyHome
+                 mozillaTunnel: {
+                    tunnel: boolean,
+                    tunnelName: string, - e.g. MyName
+                    tunnelEmail: string
+                    }
+ */
+
+SettingsController.put('/domain', async (request, response) => {
+  if (!request.body ||
+      !request.body.hasOwnProperty('local') &&
+      !request.body.local.hasOwnProperty('localDNSname') ||
+      !request.body.local.hasOwnProperty('multicastDNSstate')) {
+    response.statusMessage = 'Invalid request.';
+    response.status(400).end();
+    return;
+  }
+  try {
+    const requestContents = request.body;
+    const requestDomainName = request.body.local.localDNSname;
+    const requestState = request.body.local.multicastDNSstate;
+
+    if (requestContents.local.localDNSname) {
+      mDNSserver.server.setLocalDomain(requestDomainName);
+      await Settings.set('localDNSname', mDNSserver.server.localDomain);
+      console.log('MDNS name is: ', mDNSserver.server.localDomain);
+    }
+
+    await Settings.set('multicastDNSstate',
+                       requestState);
+    mDNSserver.server.startService(requestState);
+
+    const url = `https://${requestDomainName}.local:` +
```

Comment:
This isn't necessarily the right redirect. The user may not have https setup, so this redirect would then fail. Instead, you should check `request.secure`, which will tell you if the request came in via https. If not, you should use the http port.

Reply:
Oh! Thank you.
I've updated this.
I've also noticed a way to consolidate and improve the function which I have taken the liberty of also adding.